exec.py

```diff
@@ -124,47 +145,139 @@ def plot_plane(ws, tag):
-        rbwsfac.set_jer_source("./input/jer.root","jer_data")
+        rbwsfac.set_jer_source("./input/jer.root","jer_data",args.dummyjer)
         rbwsfac.build()
         ws = rbwsfac.get_ws()
-        ws.Print("v")
+        if do_print:
+            ws.Print("v")
 
-        # for i in range(int(ws.var("njets").getValV())):
-        #     print(i, jets[i].px, ws.var(f"gen_px_{i}").getValV(), ws.var(f"reco_px_{i}").getValV())
+        if do_plot:
+            plot_plane(ws, tag=f"{event}_before")
         
-        # for i in range(int(ws.var("njets").getValV())):
-        #     print(i, jets[i].py, ws.var(f"gen_py_{i}").getValV(), ws.var(f"reco_py_{i}").getValV())
-
-
-        f=r.TFile(f"./output/ws_{event}.root","RECREATE")
-        plot_plane(ws, tag=f"{event}_before")
-        ws.Write('before')
+        f.cd()
+        ws.Write(f'before_{event}')
         m = r.RooMinimizer(ws.function("nll"))
         m.migrad()
-        ws.Write('after')
-        plot_plane(ws, tag=f"{event}_after")
-
-    print('*'*20)
-    print('JOB INFO:')
-    endtime = time.time()
-    timeinterval = endtime - starttime
-    timeinterval_per_event = (endtime - starttime) / num_events
-    print(f'Ran over {num_events} events')
-    print(f'Total running time: {timeinterval:.3f} s')
-    print(f'Running time/event: {timeinterval_per_event:.3f} s')
-    print('*'*20)
+        ws.Write(f'rebalanced_{event}')
+        if do_plot:
+            plot_plane(ws, tag=f"{event}_after")
+
+    with open(logf, 'a') as logfile:
+        logfile.write('\n')
+        logfile.write(f'Finished job {time.ctime()}\n')
+        logfile.write('JOB INFO:\n')
+        endtime = time.time()
+        timeinterval = endtime - starttime
+        timeinterval_per_event = (endtime - starttime) / numevents
+        logfile.write(f'Ran over {numevents} events\n')
+        logfile.write(f'Total running time: {timeinterval:.3f} s\n')
+        logfile.write(f'Running time/event: {timeinterval_per_event:.3f} s\n')
 
     return ws
+
+def main():
+    args = parse_cli()
+
+    if not args.inpath:
+        raise RuntimeError('Please provide an input ROOT file.')
+
+    event_chunks = divide_into_chunks(args)
+    nchunks = len(event_chunks)
+
+    outdir = f'./output/{args.jobname}'
+    logdir = f'./output/{args.jobname}/logs'
+
+    # Do not allow to override an existing job output directory
+    if os.path.exists(outdir):
+        raise RuntimeError(f'Output directory exists: {outdir}, please remove it first.')
+
+    os.makedirs(outdir)
+    os.makedirs(logdir)
+
+    processes = []
+
+    # Save repo information for this job
+    versionfilepath = pjoin(outdir, 'version.txt')
+    with open(versionfilepath, 'w+') as f:
+        f.write(git_rev_parse() + '\n')
+        f.write(git_diff() + '\n')
+
+    # Number of parallel processes, for dry run it is automatically set to 1
+    if not args.dry:
+        ncores = args.ncores
+    else:
+        ncores = 1
+
+    for jobidx in range(ncores):
```

Comment:
I'm not familiar with this construction. Don't you want to use `multiprocessing.Pool` here?
Your code seems to indicate that the number of chunks is always identical to the number of parallel jobs to start. That implies that the chunks can get very big if the input is very big. Wouldn't it make sense to decouple the two so that you can enforce ~constant chunk SIZES rather than chunk COUNTS?

Reply:
Thanks for the suggestion, indeed ``python multiprocessing.Pool end``` works very nicely here. In 636d195, I replaced the earlier construction with the ```python Pool end``` object.